extract-hackage-info/src/Main.hs

```diff
@@ -0,0 +1,406 @@
+    concatMap injectFst
+      . Map.toList
+      . Map.map (Map.toList . Map.filter hasConflict)
+      $ hashmap
+  )
+  where
+    finalize = \case
+      [] -> unspecifiedFixityInfo
+      fs -> sconcat . NE.fromList $ fs
+    hasConflict = (> 1) . length
+    injectFst (packageName, opFixs) =
+      fmap (\(opName, fixs) -> ((packageName, opName), fixs)) opFixs
+
+extractFixitiesFromFile ::
+  -- | Hoogle database path
+  FilePath ->
+  State ->
+  -- | Current file path
+  FilePath ->
+  IO State
+extractFixitiesFromFile
+  hoogleDatabasePath
+  state@State {sProcessedFiles}
+  filePath = do
+    fileContent <- liftIO . readFileUtf8Latin1 $ filePath
+    packageName <- liftIO $ getPackageName hoogleDatabasePath filePath
+    let state' =
+          foldl' @[]
+            (onSymbolDecl packageName)
+            state
+            (fromSymbolDecl <$> symbolDecls fileContent)
+        state'' =
+          foldl' @[]
+            (onFixityDecl packageName)
+            state'
+            (fromFixityDecl <$> fixityDecls fileContent)
+        fromSymbolDecl match = capture @"declOpName" match
+        fromFixityDecl match =
+          ( capture @"fixDir" match,
+            capture @"fixPrec" match,
+            capture @"infixOpName" match
+          )
+        symbolDecls = [regex|(?m)^\s*?(?<declOpName>\([^)]+?\))\s*?::.*$|]
+        fixityDecls = [regex|(?m)^\s*?(?<fixDir>infix[rl]?)\s+?(?<fixPrec>[0-9])\s+?(?<infixOpName>[^\s]+)\s*$|]
+    return state'' {sProcessedFiles = sProcessedFiles + 1}
+
+extractHoogleInfo :: FilePath -> IO (Map String FixityMap)
+extractHoogleInfo hoogleDatabasePath = do
+  hoogleFiles <- walkDir hoogleDatabasePath (const False)
+  State {..} <-
+    foldM
+      (extractFixitiesFromFile hoogleDatabasePath)
+      initialState
+      hoogleFiles
+  putFmtLn
+    "{} hoogle files processed!"
+    (Only sProcessedFiles)
+  let (packageToOps, conflicts) = finalizePackageToOps sPackageToOps
+  putFmtLn
+    "Found {} operator declarations across {} packages for a total of \
+    \{} distinct operators"
+    (getCounts packageToOps)
+  when (not (null conflicts)) $
+    displayConflicts conflicts
+  return packageToOps
+
+displayConflicts :: [((String, String), [FixityInfo])] -> IO ()
+displayConflicts conflicts = do
+  putFmtLn
+    "Found {} conflicting declarations within packages themselves:"
+    (Only $ length conflicts)
+  TIO.putStrLn $ T.intercalate "\n" conflictLines'
+  where
+    conflictLines' = concat $ conflictLines <$> sortedConflicts
+    sortedConflicts =
+      sortBy
+        (\(packageOp1, _) (packageOp2, _) -> compare packageOp1 packageOp2)
+        conflicts
+    conflictLines ((packageName, opName), fixities) =
+      format
+        "(in {}) {}"
+        (packageName, opName)
+        : indentLines (showT <$> fixities)
+
+limitMap :: Ord k => Int -> Map k v -> Map k v
+limitMap n = Map.fromList . take n . Map.toList
+
+getCounts :: Map String FixityMap -> (Int, Int, Int)
```

Comment:
This function should have a Haddock. What counts?

Reply:
My latest commit should fix everything you listed in your latest review. Let me know if I still need to ajust something.